typedqueue/queue.go

```diff
@@ -0,0 +1,120 @@
+// Package typedqueue wraps a blobqueue, with serialization and runtime type checking of values
+package typedqueue
+
+import (
+	"encoding"
+	"fmt"
+	"reflect"
+
+	blobqueue "github.com/blueboardio/go-blobqueue"
+)
+
+type Value struct {
+	encoding.BinaryMarshaler
+	encoding.BinaryUnmarshaler
+}
+
+type Queue struct {
+	q    blobqueue.Queue
+	zero Value
+	t    reflect.Type
+}
+
+// New wraps a queue to store any value of the same type that implements
+// encoding.BinaryMarshaler and encoding.BinaryUnmarshaler.
+//
+// Any marshaling/unmarshaling error is unexpected and raises a panic.
+// So any error returned by the queue methods come from the original queue.
+func New(q blobqueue.Queue, zeroValue encoding.BinaryMarshaler) *Queue {
+	t := reflect.TypeOf(zeroValue)
+	// Checks of the contract for values
+	_ = decodeValue(t, encodeValue(t, zeroValue))
+
+	return &Queue{q: q, t: t}
+}
+
+// Len returns the length of the queue.
+func (q *Queue) Len() (int, error) {
+	return q.q.Len()
+}
+
+// Empty clears the queue.
+func (q *Queue) Empty() error {
+	return q.q.Empty()
+}
+
+func encodeValue(t reflect.Type, v encoding.BinaryMarshaler) []byte {
+	// Runtime check of type
+	// This could be removed from production builds
+	if reflect.TypeOf(v) != t {
+		panic(fmt.Errorf("unexpected value of type %T", v))
+	}
+
+	b, err := v.MarshalBinary()
+	if err != nil {
+		panic(err)
+	}
+	return b
+}
+
+func decodeValue(t reflect.Type, b []byte) interface{} {
+	v := reflect.New(t)
+	if err := v.Interface().(encoding.BinaryUnmarshaler).UnmarshalBinary(b); err != nil {
+		panic(err)
+	}
+	return v.Elem().Interface()
+}
+
+func decodeSlice(t reflect.Type, src [][]byte) interface{} {
+	if len(src) == 0 {
+		// nil slice
+		return reflect.New(reflect.SliceOf(t)).Elem().Interface()
+	}
+	sl := reflect.MakeSlice(reflect.SliceOf(t), len(src), len(src))
```

Comment:
As I understand `reflect.MakeSlice` the capacity defines the first slice allocations and len precise how many items to initialize to 0, as the function will run through the slice and set their values, is not useless to set len ?
Could we call it with a 0 len or will it cause panic issues due to nil dereference ?
```suggestion
	sl := reflect.MakeSlice(reflect.SliceOf(t), 0, len(src))
```

Reply:
If the initial len was set as 0, we would have to `append` each value.

Instead I choose to directly allocate a slice with the final length, and replace the initial zero value with the unmarshaled value. The replacement is done when the `UnmarshalBinary` is called.

Reply:
Yeah but if len is the one wich allocate, what's the purpose of capacity ?

Reply:
Capacity is the memory allocated. Len is the part of that memory effectively used. If `cap` > `len`, we have growth capacity for `append` without reallocation.
https://blog.golang.org/slices